### ISSUE TITLE
Sending and Receiving message to same topic not working

### ISSUE THREAD

**Issue by hasmqtt** [knolleary/pubsubclient#197]
Hi All,

I am trying to send and receive message to same topic, but it is not working. I read this library uses same buffer.

Is there workaround or solution available ?

Appreciate your help.

Regards

**Comment by knolleary** [comment]
It is perfectly possible to publish and subscribe to the same topic. Without any specifics it is hard to help.

**Comment by hasmqtt** [comment]
Thanks for quick response, in my code I am publishing and subscribing to same topic. I am able to receive the messages published by another client (MQTT SPY) to ESP8266 but not able to publish anything from ESP8266. Here is code please change WiFi details

#include <ESP8266WiFi.h>
#include <PubSubClient.h>

// Update these with values suitable for your network.

const char* ssid = "......................";
const char* password = "...................";
const char* mqtt_server = "m10.cloudmqtt.com";
int mqtt_port = 12438;
const char* mqtt_user = "test1";
const char* mqtt_password = "test1234";
const char* topic1 = "PSN/22104";
void callback(char* topic, byte* payload, unsigned int length);

const int ID = ESP.getChipId();
const String mqtt_device_name = "ESP8266-" + String(ID);

//WiFiClient espClient;
//PubSubClient client(mqtt_server, mqtt_port, callback, espClient);

WiFiClient espClient;
PubSubClient mqttClient(mqtt_server, mqtt_port, callback, espClient);


long lastMsg = 0;
char msg[50];
int value = 0;

void setup_wifi() {

  delay(10);
  // We start by connecting to a WiFi network
  Serial.println();
  Serial.print("Connecting to ");
  Serial.println(ssid);

  WiFi.begin(ssid, password);

  while (WiFi.status() != WL_CONNECTED) {
    delay(500);
    Serial.print(".");
  }

  randomSeed(micros());

  Serial.println("");
  Serial.println("WiFi connected");
  Serial.println("IP address: ");
  Serial.println(WiFi.localIP());
}

void callback(char* topic, byte* payload, unsigned int length) {
  Serial.print("Message arrived [");
  Serial.print(topic);
  Serial.print("] ");
  for (int i = 0; i < length; i++) {
    Serial.print((char)payload[i]);
  }
  Serial.println();

  // Switch on the LED if an 1 was received as first character
  if ((char)payload[0] == '1') {
    digitalWrite(BUILTIN_LED, LOW);   // Turn the LED on (Note that LOW is the voltage level
    // but actually the LED is on; this is because
    // it is acive low on the ESP-01)
  } else {
    digitalWrite(BUILTIN_LED, HIGH);  // Turn the LED off by making the voltage HIGH
  }

}

void mqttConnectAndSubscribe() {
  if(WiFi.isConnected()) {
    Serial.print("Checking MQTT connection...");
    if(!mqttClient.connected()) {
      Serial.print("MQTT Not connected, trying to connect, ");
      // Attempt to connect
      mqttClient.connect(mqtt_device_name.c_str(), mqtt_user, mqtt_password);
      if (mqttClient.connected()) {
        Serial.println("connected and subscribed to topic");
        mqttClient.subscribe(topic1);
      }
    } else {
      Serial.println("failed.");
    }
  } else {
    Serial.println("Not connected to AP, so not publishing to MQTT broker");
  }
}

void publishStatus() {
  if (mqttClient.connected()) {
    Serial.print("MQTT connected, ");
    mqttClient.publish(topic1, "gpioStatus().c_str()");
    Serial.println(" Published Status message");
  } else {
    Serial.println("MQTT connect failed.");
  }
}

void setup() {
  pinMode(BUILTIN_LED, OUTPUT);     // Initialize the BUILTIN_LED pin as an output
  Serial.begin(115200);
  setup_wifi();
  mqttConnectAndSubscribe();
}

void loop() {

  if (!mqttClient.connected()) {
    mqttConnectAndSubscribe();
  }
  mqttClient.loop();

  long now = millis();
  if (now - lastMsg > 2000) {
    lastMsg = now;
    ++value;
    snprintf (msg, 75, "hello world #%ld", value);
    Serial.print("Publish message: ");
    Serial.println(msg);
    mqttClient.publish("PSN/22104/Hall", msg);
  }
}

**Comment by knolleary** [comment]
What about it isn't working?

**Comment by knolleary** [comment]
I can see a couple of different publish calls - one in `publishStatus` to the topic defined in `topic1` (which is the topic you subscribe to) and the other to `PSN/22104/Hall` in the main `loop` function.

`publishStatus` is never called, so the only attempted publish will be the one in the main `loop`. So subscribing/publishing to the same topic is not involved here.

Small aside - the `snprintf`call says the max size for `msg` is 75, but in its original declaration, `msg` is only 50 chars.

So what do you see in the serial output? Do you see the `Publish message: X` output to confirm it is calling the publish function?

**Comment by hasmqtt** [comment]
Thanks that was silly mistake I did, never called publishStatus from loop, actually this test code is extracted from my other code. 

It is working beautifully without any problem, actual problem was buffer size in original code once I increased MQTT_MAX_PACKET_SIZE 256 it started working like charm.

One more small question, if I define MQTT_MAX_PACKET_SIZE in my code it should take but  it is not working I have to edit PubSubClient.h then it works. I am new to C++ macro so not sure what is the issue.
